Objects/typeobject.c

```diff
@@ -3078,19 +3079,28 @@ type_new_descriptors(const type_new_ctx *ctx, PyTypeObject *type)
         }
     }
 
-    if (ctx->add_dict && ctx->base->tp_itemsize) {
-        type->tp_dictoffset = -(long)sizeof(PyObject *);
-        slotoffset += sizeof(PyObject *);
+    if (ctx->add_dict) {
+        if (ctx->base->tp_itemsize) {
+            type->tp_dictoffset = -(long)sizeof(PyObject *);
+            slotoffset += sizeof(PyObject *);
+        }
+        else {
+            assert((type->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0);
+            type->tp_flags |= Py_TPFLAGS_MANAGED_DICT;
+        }
     }
 
     if (ctx->add_weak) {
         assert(!ctx->base->tp_itemsize);
-        type->tp_weaklistoffset = slotoffset;
-        slotoffset += sizeof(PyObject *);
+        if (type->tp_flags & Py_TPFLAGS_MANAGED_DICT) {
+            type->tp_weaklistoffset = -4 * sizeof(PyObject *);
```

Comment:
Is the pre-header layout documented, either in a comment, the docs, or an adjacent .txt file?  Either way, I don't see the weaklist's place in the pre-header added to such documentation in this PR.  That information would be super helpful, since I can only guess what `-4 * sizeof(PyObject *)` means.  A short comment here pointing to that documentation would be nice too.

Reply:
Documenting the object layout is on the to-do list.
Early draft here: https://github.com/faster-cpython/cpython/blob/document-object-layout/Objects/object_layout.md

Reply:
What?  No SVG? <wink>